File: site/tools/generate-docs/package.ts
import { Daemon } from './daemon';
import Enum from './enum';
import { Message } from './message';
import { Service } from './service';
import { JsonProtoFile } from './types';

const { log } = console;

export class Package {
  daemon: Daemon;

  name: string;
  description: string;
  messages = new Map<string, Message>();
  enums = new Map<string, Enum>();
  services: Service[] = [];
  experimental = false;

  constructor(name: string, daemon: Daemon) {
    log(`Creating package ${name}`);
    this.daemon = daemon;
    this.name = name;
  }

  addProtoFile(json: JsonProtoFile) {
    log(`Adding proto file ${json.name} to package ${json.package}`);
    this.name = json.package;
    this.description = json.description;

    log(`Adding ${json.messages.length} messages in ${json.package}`);
    json.messages.forEach((m) =>
      this.messages.set(m.longName, new Message(m, this, json.name))
    );

    log(`Adding ${json.enums.length} enums in ${json.package}`);
    json.enums.forEach((e) =>
      this.enums.set(e.longName, new Enum(e, this, json.name))
    );

    log(`Adding ${json.services.length} services in ${json.package}`);
    json.services.forEach((s) =>
      this.services.push(new Service(s, this, json.name))
    );
  }

  exportMarkdown(daemonName: string) {
    log(`Exporting package ${this.name}`);
    this.services.forEach((s) => s.exportMarkdown(daemonName));
  }
}

File: models/package.go
package models

import (
	"fmt"

	"github.com/lightninglabs/lightning-api-ng/defs"
)

type Package struct {
	Name         string
	Description  string
	Experimental bool
	Messages     map[string]*Message
	Services     []*Service

	App *App
}

// NewPackage creates a new package model from a package definition
func NewPackage(name string, app *App) *Package {
	fmt.Printf("Creating package %s\n", name)
	return &Package{
		Name:     name,
		App:      app,
		Messages: make(map[string]*Message),
		Services: make([]*Service, 0),
	}
}

// AddProtoFile adds the messages, enums, and services from the given proto
// file definition to the package.
func (p *Package) addProtoFile(file *defs.File) {
	fmt.Printf("Adding proto file %s to package %s\n", file.Name,
		file.Package)

	fmt.Printf("Adding %d messages to package %s\n", len(file.Messages),
		file.Package)
	for _, msgDef := range file.Messages {
		msg := NewMessage(msgDef, p, file.Name)
		p.Messages[msg.LongName] = msg
	}

	fmt.Printf("Adding %d services to package %s\n", len(file.Services),
		file.Package)
	for _, svcDef := range file.Services {
		svc := NewService(svcDef, p, file.Name)
		p.Services = append(p.Services, svc)
	}
}

// ExportMarkdown exports the services in this package to markdown.
func (p *Package) ExportMarkdown() error {
	fmt.Printf("Exporting package %s\n", p.Name)
	for _, svc := range p.Services {
		err := svc.ExportMarkdown()
		if err != nil {
			return err
		}
	}
	return nil
}
